### PR TITLE
Use async for CLI commands

### PR DESCRIPTION
**Pull request by GermanBluefox**
In this PR some of the most complex CLI commands were refactored to use asyc/await.

### CODE REVIEW

**Comment by GermanBluefox** [comment]
I suggest to merge it before https://github.com/ioBroker/ioBroker.js-controller/pull/1497

File: README.md [hunk 1]
```diff
@@ -23,24 +23,16 @@ To upgrade your Node.js version and ioBroker please follow https://forum.iobroke
...
 
 ----------------------------------------------------------------------
 
 ## Usage
-
-### Install `Node.js`
-
-Example for a Debian based system:
-
+Example hwo to install on a Debian based system:
```
Comment (foxriver76) [inline README.md:33]: typo

File: README.md [hunk 2]
```diff
@@ -143,13 +135,13 @@ If you want to disable the error reporting you can do this by setting the state
 ### Notification System
 **Feature status:** Technology preview since js-controller 3.2.0
 
-The notification system in ioBroker allows to set, detect and store notifications per Host and allows to query the details.
+The notification system in ioBroker allows setting, detect and store notifications per Host and allows to query the details.
```
Comment (foxriver76) [inline README.md:138]: then we would also need to change to `detecting and storing`

File: README.md [hunk 3]
```diff
@@ -190,7 +182,7 @@ All three are optional and can be a string or null/undefined if ommited.
 #### Log levels
 **Feature status:** stable
 
-The js-controller and each adapter has defined it's own log level. By default, `info` is used. The following loglevels can be used:
+The js-controller and each adapter has defined its own log level. By default, `info` is used. The following log-levels can be used:
```
Comment (foxriver76) [inline README.md:185]: better with a space between?

File: README.md [hunk 4]
```diff
@@ -217,7 +209,7 @@ The log level can be changed dynamically for adapter-instance and host (main con
 
 The states `system.adapter.xy.logLevel` and `system.host.hostname.logLevel` are updated on instance/controller start with the configured log level and can afterwards be used to change the loglevel during runtime. These changes are __not__ persisted, so the next restarts resets the loglevel to the configured one.
 
-This possibility allows to better debug adapters also during runtime.
+This possibility allows to debug adapters better also during runtime.
```
Comment (foxriver76) [inline README.md:212]: imo this is more confusing than before

File: README.md [hunk 5]
```diff
@@ -519,7 +511,7 @@ The following fields are allowed in the alias structure:
     * `alias.id.write` contains the ID of the object which will be set when alias is written
     * `alias.id.read` contains the ID of the object which will be mirrored to the alias object/state   
 * `alias.read` can optionally contain a read script (will be evaluated) to calculate the alias value when the target state changes
-* `alias.write` can optionally contain a write script (will be evaluated) to calculate the target value if the alias value is changed
+* `alias.write` can optionally contain a write-script (will be evaluated) to calculate the target value if the alias value is changed
```
Comment (foxriver76) [inline README.md:514]: read script stays the same and write-script gets a hypen, does not make sense, would keep it like before

File: packages/common/lib/common/tools.js [hunk 6]
```diff
@@ -2450,51 +2503,49 @@ function validateGeneralObjectProperties(obj, extend) {
  * @alias getAllInstances
  * @memberof tools
  * @param {string[]} adapters list of adapter names to get instances for
+ * @param {object} objects class redis objects
  * @param {function} callback callback to be executed
  */
 function getAllInstances(adapters, objects, callback) {
+    showDeprecatedMessage('tools.getAllInstances');
```
Comment (foxriver76) [inline packages/common/lib/common/tools.js:2510]: Do we really need this, is someone from outside accessing js-controller tools, else we should just fix all occurences.

File: packages/common/lib/common/tools.js [hunk 7]
```diff
@@ -3095,6 +3146,11 @@ async function setExecutableCapabilities(execPath, capabilities, modeEffective,
     }
 }
 
+function showDeprecatedMessage(func, logger) {
```
Comment (foxriver76) [inline packages/common/lib/common/tools.js:3149]: as mentioned above I would assume we do not need it

File: packages/common/lib/common/tools.js [hunk 8]
```diff
@@ -649,6 +649,59 @@ function getJson(urlOrPath, agent, callback) {
     }
 }
 
+async function getJsonAsync(urlOrPath, agent) {
```
Comment (foxriver76) [inline packages/common/lib/common/tools.js:652]: for new methods some jsdoc would be really nice

File: packages/controller/lib/setup/setupRepo.js [hunk 9]
```diff
@@ -1,14 +1,13 @@
...
-const fs = require('fs');
-
 /** @class */
 function Repo(options) {
     const { EXIT_CODES } = require('@iobroker/js-controller-common');
     const { tools }  = require('@iobroker/js-controller-common');
     const axios      = require('axios');
     const ioPackage  = require('../../io-package.json');
     const version    = ioPackage.common.version;
+    const fs         = require('fs');
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:10]: guess we also should prefer fs-extra here

File: packages/controller/lib/setup/setupRepo.js [hunk 10]
```diff
@@ -195,24 +194,13 @@ function Repo(options) {
             }
 
             // update variables of every admin instance
-            await new Promise(resolve => updateInfo(allSources, () => resolve()));
+            await updateInfo(allSources);
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:197]: should be wrapped in try catch afais

File: packages/controller/lib/setup/setupRepo.js [hunk 11]
```diff
@@ -227,154 +215,115 @@ function Repo(options) {
...
-                    link: repoUrl,
-                    json: null
-                };
-                obj.from = 'system.host.' + tools.getHostName() + '.cli';
-                obj.ts = Date.now();
-                objects.setObject('system.repositories', obj, callback);
-            }
-        });
+        if (obj.native.repositories[repoName]) {
+            return `Repository "${repoName}" yet exists: ${obj.native.repositories[repoName].link}`;
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:265]: prevously this was first cb arg so probably an error argument, should it not be a throw here then?

File: packages/controller/lib/setup/setupRepo.js [hunk 12]
```diff
@@ -227,154 +215,115 @@ function Repo(options) {
...
-                callback && callback(err);
-            } else {
-                if ((obj.common.activeRepo && typeof obj.common.activeRepo === 'string' && obj.common.activeRepo === repoName) ||
-                    (obj.common.activeRepo && typeof obj.common.activeRepo === 'object' && obj.common.activeRepo.includes(repoName))) {
-                    callback && callback(`Cannot delete active repository: ${repoName}`);
+    this.del = async function (repoName) {
+        const obj = await objects.getObjectAsync('system.config');
+        if ((obj.common.activeRepo && typeof obj.common.activeRepo === 'string' && obj.common.activeRepo === repoName) ||
+            (obj.common.activeRepo && typeof obj.common.activeRepo === 'object' && obj.common.activeRepo.includes(repoName))) {
+            return `Cannot delete active repository: ${repoName}`;
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:281]: same error cb?

File: packages/controller/lib/setup/setupRepo.js [hunk 13]
```diff
@@ -227,154 +215,115 @@ function Repo(options) {
...
     };
 
-    this.setActive = function (repoName, callback) {
-        objects.getObject('system.repositories', (err, obj) => {
-            if (err) {
-                return callback && callback(err);
-            } else {
-                obj = obj || defaultSystemRepo;
+    this.setActive = async function (repoName) {
+        const obj = (await objects.getObjectAsync('system.repositories')) || defaultSystemRepo;
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:298]: I personally find that style strange would prefer two lines

File: packages/controller/lib/setup/setupRepo.js [hunk 14]
```diff
@@ -227,154 +215,115 @@ function Repo(options) {
...
+        if (!obj.native.repositories[repoName]) {
+            return `Repository "${repoName}" not found.`;
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupRepo.js:300]: error cb would let it throw

**Review by foxriver76 — CHANGES_REQUESTED** [review]
Okay so I've reviewed first 9 files. I cannot make it in one session and cannot save my review as a draft so here is a partial review :-)

File: packages/common/lib/common/tools.js (hunk 6 above)
Comment (GermanBluefox) [reply packages/common/lib/common/tools.js:2510]: It is for me. Later when I could see no one uses getAllInstances, I will remove it all. It is hard to change ALL functions on ALL places at once.

File: packages/controller/lib/setup/setupRepo.js (hunk 9 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupRepo.js:10]: ONly existsSync, readFileSync are used. Why use fs-extra here?

File: packages/common/lib/common/tools.js (hunk 8 above)
Comment (GermanBluefox) [reply packages/common/lib/common/tools.js:658]: added

File: packages/controller/lib/setup/setupRepo.js (hunk 10 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupRepo.js]: Done

File: README.md (hunk 4 above)
Comment (GermanBluefox) [reply README.md]: This possibility allows to debug adapters better and during runtime.

File: README.md (hunk 2 above)
Comment (GermanBluefox) [reply README.md]: The notification system in ioBroker allows setting, detecting and storing notifications ...

File: packages/common/lib/common/tools.js (hunk 7 above)
Comment (GermanBluefox) [reply packages/common/lib/common/tools.js:3156]: Will be deleted later. When all CLI command  will be converted.

File: packages/controller/lib/setup/setupUpload.js [hunk 15]
```diff
@@ -743,64 +710,67 @@ function Upload(options) {
...
             // Always update installed From from File on disk if exists and set
-            if (iopackFile && iopackFile.common && iopackFile.common.installedFrom) {
-                iopack.common = iopack.common || {};
-                iopack.common.installedFrom = iopackFile.common.installedFrom;
+            if (ioPackFile && ioPackFile.common && ioPackFile.common.installedFrom) {
+                ioPack.common = ioPack.common || {};
+                ioPack.common.installedFrom = ioPackFile.common.installedFrom;
+            }
+            // Not existing? Why ever ... we recreate
+            const obj = (await objects.getObject('system.adapter.' + name)) || {};
```
Comment (foxriver76) [inline packages/controller/lib/setup/setupUpload.js:745]: I am not a fan of that, at the latest if we need to wrap try catch around it we would need to split, would just make two lines out of it

File: packages/controller/lib/setup.js [hunk 16]
```diff
@@ -783,25 +771,39 @@ async function processCommand(command, args, params, callback) {
                                 adapters.push(objs.rows[i].value.common.name);
                             }
 
-                            upload.uploadAdapterFull(adapters, callback);
-                        });
+                            await upload.uploadAdapterFullAsync(adapters);
+                            callback();
+                        } catch (err) {
+                            console.error(`Cannot upload all adapters: ${err}`);
```
Comment (foxriver76) [inline packages/controller/lib/setup.js:777]: `e.message` here and for all others in setup?

**Review by foxriver76 — CHANGES_REQUESTED** [review]
Ok, I hope I've got all now, mostly just small things. But we should really think about the `tools.maybeCallback` vs deprecation log approach.

File: packages/controller/lib/setup.js (hunk 16 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup.js]: All e => err

File: packages/controller/lib/setup/setupUpgrade.js [hunk 17]
```diff
@@ -263,56 +261,53 @@ function Upgrade(options) {
      * @param {boolean} upgradeAll if true, this is an upgrade all call, we don't do major upgrades if no tty
      * @param {function} callback callback to be executed after function is finished
      */
-    this.upgradeAdapter = async function (repoUrl, adapter, forceDowngrade, autoConfirm, upgradeAll, callback) {
+    this.upgradeAdapter = (repoUrl, adapter, forceDowngrade, autoConfirm, upgradeAll, callback) => {
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupUpgrade.js:264]: The Idea is to change all Async into normal one later, as the functions are not used outside of controller. 
But it is hard to change ALL functions in all blocks at once. So one after each other.

[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/controller/lib/setup/setupUpgrade.js:264]: Yes and this is exactly the reason why we added `maybeCallback` to slowly move methods from cb style to promise style without breaking anything. Opinions @Apollon77 @AlCalzone

File: packages/controller/lib/setup/setupBackup.js [hunk 18]
```diff
@@ -55,55 +55,45 @@ class BackupRestore {
...
-                            }
-                        });
+                        await this._copyFile(id, `${srcPath}/${res[t].file}`, `${destPath}/${res[t].file}`);
                     }
                 }
             }
-            if (!count) {
-                return tools.maybeCallback(callback);
+        } catch (err) {
+            if (!err.toString().includes('Not exists')) {
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupBackup.js:93]: err alsowhen "not exists" should only be Error object from my rememberings ...

File: packages/controller/lib/setup/setupInstall.js [hunk 19]
```diff
@@ -640,266 +613,235 @@ function Install(options) {
...
 
-        objects.getObject('system.adapter.' + adapter, (err, doc) => {
-            // Adapter is not installed - install it now
-            if (err || !doc || !doc.common.installedVersion) {
-                return that.installAdapter(adapter, () =>
-                    that.createInstance(adapter, options, callback));
-            }
+        let doc = await objects.getObjectAsync('system.adapter.' + adapter);
+        // Adapter is not installed - install it now
+        if (!doc || !doc.common.installedVersion) {
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupInstall.js:618]: original code also ignored the error case and executed the code from this if

File: packages/controller/lib/setup/setupInstall.js [hunk 20]
```diff
@@ -640,266 +613,235 @@ function Install(options) {
...
-                                processExit(EXIT_CODES.NO_MULTIPLE_INSTANCES_ALLOWED);
-                                return;
-                            }
+        const res = await objects.getObjectViewAsync('system', 'instance', {
+            startkey: 'system.adapter.' + adapter + '.',
+            endkey: 'system.adapter.' + adapter + '.\u9999'
+        });
+        const systemConfig = await objects.getObjectAsync('system.config');
+        const defaultLogLevel = systemConfig && systemConfig.common && systemConfig.common.defaultLogLevel;
+        let a;
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupInstall.js:633]: Are not block scoped variabled better ? ;-) Just used in some for loops or?

File: packages/controller/lib/setup/setupRepo.js [hunk 21]
```diff
@@ -227,154 +219,117 @@ function Repo(options) {
...
-                                activeRepo = [activeRepo];
-                            }
-                            console.log(`\nActive repo(s): ${activeRepo.join(', ')}`);
-                        }
-                        callback();
-                    });
-                } else {
-                    console.error('Cannot get list: ' + err);
-                    callback(EXIT_CODES.CANNOT_GET_REPO_LIST);
+    this.showRepoStatus = async function () {
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupRepo.js:235]: Really with "function"? Also cases below. Yes, no "this" is used but still :-)

File: packages/controller/lib/setup/setupUpgrade.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/controller/lib/setup/setupUpgrade.js:267]: We agreed that the deprecations will be in master during the rest of the work to convert CLI stuff to async and will be gone before we release 4.0

File: packages/controller/lib/setup/setupUpgrade.js [hunk 22]
```diff
@@ -468,160 +451,149 @@ function Upgrade(options) {
...
                     }
                 } catch (err) {
                     console.log(`Can not check version information to display upgrade infos: ${err.message}`);
                 }
 
                 console.log(`Update ${adapter} from @${ioInstalled.common.version} to @${targetVersion}`);
                 // Get the adapter from web site
-                install.downloadPacket(sources, `${adapter}@${targetVersion}`, null, (enableAdapterCallback, name, ioPack) =>
-                    finishUpgrade(name, ioPack, () => enableAdapterCallback(callback)));
+                const name = await install.downloadPacketAsync(sources, `${adapter}@${targetVersion}`);
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupUpgrade.js:474]: Is the "enableAdapterCallback logic" gone here?

File: packages/controller/lib/setup/setupUpload.js [hunk 23]
```diff
@@ -464,343 +421,345 @@ function Upload(options) {
...
 
-                obj.from = `system.host.${hostname}.cli`;
-                obj.ts = Date.now();
+            obj.from = `system.host.${hostname}.cli`;
+            obj.ts = Date.now();
 
-                objects.setObject('system.adapter.' + name, obj, () =>
-                    this._upgradeAdapterObjectsHelper(name, iopack, hostname, logger, callback));
-            });
+            await objects.setObjectAsync('system.adapter.' + name, obj);
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupUpload.js:758]: error here was ignored in old code, now would prevent the  next call

File: packages/controller/lib/setup/setupUpload.js [hunk 24]
```diff
@@ -464,343 +421,345 @@ function Upload(options) {
...
             // Always update installed From from File on disk if exists and set
-            if (iopackFile && iopackFile.common && iopackFile.common.installedFrom) {
-                iopack.common = iopack.common || {};
-                iopack.common.installedFrom = iopackFile.common.installedFrom;
+            if (ioPackFile && ioPackFile.common && ioPackFile.common.installedFrom) {
+                ioPack.common = ioPack.common || {};
+                ioPack.common.installedFrom = ioPackFile.common.installedFrom;
+            }
+            // Not existing? Why ever ... we recreate
+            let obj = await objects.getObject('system.adapter.' + name);
```
Comment (Apollon77) [inline packages/controller/lib/setup/setupUpload.js:733]: try/catch? exception was handled earlier like empty object ... intended?

**Review by Apollon77 — COMMENTED** [review]
some stuff to check I think.

Main topic are in my eyes:
* why you added all these err.messager || err? All thrown Error should be Error objects and with just outputtin err.message We will see each case where something else happens ... with this fallback we will never see where we missed places that do not throw Error objects
* when we look in exceptions we should log err.message to also show sentrythat we handled it ( I would like to add sentry to CLI also before 4.0 release because it is a blind spot we have)

And some places where this errors are handled now differently and also sthe "reenable adapters after an update/install feeld now not longer working (unless I oversee something)

File: packages/controller/lib/setup.js [hunk 25]
```diff
@@ -721,20 +717,14 @@ async function processCommand(command, args, params, callback) {
...
-                                    install.createInstance(name, params, callback);
-                                } else {
-                                    return void callback();
-                                }
-                            });
-                        });
-                    });
+                    await install.downloadPacketAsync(repoUrl, installName);
+                    await install.installAdapterAsync(installName, repoUrl);
+                    await install.createInstance(name, params, callback);
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js]: enableAdapterCallback  is obsolete

File: packages/controller/lib/setup.js [hunk 26]
```diff
@@ -956,11 +958,17 @@ async function processCommand(command, args, params, callback) {
...
+                        if (adapter === 'self') {
+                            const hostAlive = await states.getStateAsync(`system.host.${tools.getHostName()}.alive`);
+                            await upgrade.upgradeControllerAsync('', params.force || params.f, hostAlive && hostAlive.val);
+                        } else {
+                            await upgrade.upgradeAdapterAsync('', adapter, params.force || params.f, params.y || params.yes, false);
+                        }
+                        return void callback();
+                    } catch (err) {
+                        console.error(`Cannot upgrade: ${err.message || err}`);
+                        return void callback(EXIT_CODES.INVALID_REPO);
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js:983]: Yes. INVALID_REPO means something wrong with the repo or with the loading of repo

File: packages/controller/lib/setup.js [hunk 27]
```diff
@@ -2360,78 +2375,75 @@ async function processExit(exitCode) {
...
-                console.warn(`[Not critical] Cannot delete object ${id}: ${JSON.stringify(err)}`);
-            }
-            setImmediate(delObjects, ids, callback);
-        });
+const EXCEPTIONS = [
+    '0_userdata.0',
+    'alias.0',
+    'enum.functions',
+    'enum.rooms',
+    'system.config',
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js:2396]: system.config may not be deleted

File: packages/controller/lib/setup/setupInstall.js (hunk 20 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupInstall.js:633]: It is old code

File: packages/controller/lib/setup/setupRepo.js (hunk 21 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupRepo.js:235]: this still work with such a construction. And it is all old stuff, that I did't change

File: packages/controller/lib/setup/setupUpgrade.js (hunk 22 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupUpgrade.js:479]: It is all now on one place in setupInstall.js:
image

File: packages/adapter/lib/adapter/adapter.js [hunk 28]
```diff
@@ -49,7 +49,8 @@ const supportedFeatures = [
     'ADAPTER_AUTO_DECRYPT_NATIVE', // all native attributes, that are listed in an array `encryptedNative` in io-pack will be automatically decrypted and encrypted. Since js-controller 3.0
     'PLUGINS', // configurable plugins supported. Since js-controller 3.0
     'CONTROLLER_NPM_AUTO_REBUILD', // Automatic rebuild when node version mismatch is detected. Since js-controller 3.0
-    'CONTROLLER_READWRITE_BASE_SETTINGS' // If base settings could be read and written. Since js-controller 3.0
+    'CONTROLLER_READWRITE_BASE_SETTINGS', // If base settings could be read and written. Since js-controller 3.0
+    'CONTROLLER_MULTI_REPO' // Controller supports multiple repositories
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/adapter/lib/adapter/adapter.js:54]: Done

File: packages/common/lib/common/tools.js [hunk 29]
```diff
@@ -649,6 +649,65 @@ function getJson(urlOrPath, agent, callback) {
...
+        if (urlOrPath.startsWith('http://') || urlOrPath.startsWith('https://')) {
+            try {
+                const result = await axios(urlOrPath, {timeout: 10000, headers: {'User-Agent': agent}, validateStatus: status => status !== 200});
+                return result.data;
+            } catch (error) {
+                console.warn(`Cannot download json from ${urlOrPath}. Error: ${error}`);
+                return null;
+            }
+        } else {
+            if (fs.existsSync(urlOrPath)) {
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/common/lib/common/tools.js:675]: Done

File: packages/controller/lib/setup.js [hunk 30]
```diff
@@ -530,16 +529,13 @@ async function processCommand(command, args, params, callback) {
...
-                        });
+                        const repo = new Repo({objects, states});
 
                         try {
                             await repo.rename('default', 'stable', 'http://download.iobroker.net/sources-dist.json');
                             await repo.rename('latest', 'beta', 'http://download.iobroker.net/sources-dist-latest.json');
-                        } catch (e) {
-                            console.warn(e.message);
+                        } catch (err) {
+                            console.warn(`Cannot rename: ${err.message || err}`);
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js:531]: Changed

File: packages/controller/lib/setup.js [hunk 31]
```diff
@@ -2360,78 +2375,75 @@ async function processExit(exitCode) {
...
-            if (err &&
-                id !== 'system.group.user' &&
-                id !== 'system.group.administrator' &&
-                id !== 'system.user.admin'
-            ) {
-                console.warn(`[Not critical] Cannot delete object ${id}: ${JSON.stringify(err)}`);
-            }
-            setImmediate(delObjects, ids, callback);
-        });
+const EXCEPTIONS = [
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js]: OBJECTS_THAT_CANNOT_BE_DELETED

File: packages/controller/lib/setup.js [hunk 32]
```diff
@@ -2360,78 +2375,75 @@ async function processExit(exitCode) {
...
 }
 
-function cleanDatabase(isDeleteDb, callback) {
-    let taskCnt = 0;
+async function delStates() {
+    const keys = await states.getKeys('*');
+    if (keys) {
+        console.log(`clean ${keys.length} states...`);
+        for (let i = 0; i < keys.length; i++) {
+            await states.delState(keys[i]);
```
[the remark this answers is not in the document]
Comment (GermanBluefox) [reply packages/controller/lib/setup.js]: Added

File: packages/controller/lib/setup/setupBackup.js (hunk 18 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupBackup.js]: err.message used

File: packages/controller/lib/setup/setupInstall.js (hunk 19 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupInstall.js]: Added err processing

File: packages/controller/lib/setup/setupInstall.js (hunk 20 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupInstall.js]: Changed

File: packages/controller/lib/setup/setupUpload.js (hunk 23 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupUpload.js]: Added try/catch

File: packages/controller/lib/setup/setupUpload.js (hunk 24 above)
Comment (GermanBluefox) [reply packages/controller/lib/setup/setupUpload.js:742]: Added try/catch

**Comment by Apollon77** [comment]
@GermanBluefox It seems that merging the one "silly logPR" got a conflict again ... and also testing fails

https://github.com/ioBroker/ioBroker.js-controller/runs/4423586051?check_suite_focus=true#step:8:379

Pleas echeck

**Pull request merged**